tests/accelerator/sycl/CMakeLists.txt

```diff
@@ -19,11 +19,23 @@ target_link_libraries( algebra_tests_sycl_common
              vecmem::core vecmem::sycl )
 add_library( algebra::tests_sycl_common ALIAS algebra_tests_sycl_common )
 
-# Set up all of the (available) CUDA tests.
+# Set up all of the (available) SYCL tests.
 algebra_add_test( array_sycl "array_cmath.sycl"
    LINK_LIBRARIES algebra::array_cmath algebra::tests_sycl_common
                   GTest::gtest_main )
 
+if( ALGEBRA_PLUGINS_INCLUDE_EIGEN )
+   algebra_add_test( eigen_sycl "eigen_cmath.sycl" "eigen_eigen.sycl"
+      LINK_LIBRARIES algebra::eigen_cmath algebra::eigen_eigen
+                     algebra::tests_sycl_common GTest::gtest_main )
+   # Prevent Eigen from considering the SYCL device compilation as either:
+   #  - host compilation;
+   #  - CUDA compilation;
+   #  - HIP compilation.
+   target_compile_definitions( algebra_test_eigen_sycl
+      PRIVATE EIGEN_NO_CUDA EIGEN_NO_HIP EIGEN_DONT_VECTORIZE )
```

Comment:
Does "NO CUDA compilation" imply that we are not supposed to use SYCL with cuda backend?

Reply:
Quite the opposite. :smile:

When building this test with the Intel compiler's CUDA backend, the preprocessor macros in Eigen think that they are being processed by `nvcc`. And try to make use of symbols that LLVM doesn't know about.

With `EIGEN_NO_CUDA` we tell Eigen not to use CUDA specific formalisms, and just rely on standard C\+\+.

`EIGEN_DON_VECTORIZE` is very similar. If one compiles for the `spir64_x86_64` backend at the same time as for any of the GPU backends, Eigen tries to use vectorization intrinsics in the code. Which of course the device code compiler freaks out about.

So yeah, Eigen still has ways to go for convenient SYCL usage. :stuck_out_tongue: (Though the phase space with this is so huge that I wonder if it even makes sense to try to automate all of this behind the scenes...)

Reply:
P.S. I am able to run these tests with the latest "Intel compiler nightly" from a single test binary on OpenCL, CUDA and HIP backends at the same time. :wink: